### PR TITLE
[deprecation fix] server ip is deprecated for GET|POST /reset/

### PR DESCRIPTION
**Pull request by acjohnson**

### CODE REVIEW

File: hetzner/reset.py [hunk 1]
```diff
@@ -83,9 +83,14 @@ def observed_reboot(self, patience=300, tries=None, manual=False):
...
-            self.server.logger.info("Tring to reboot using the %r method.",
+            self.server.logger.info("Trying to reboot using the %r method.",
                                     mode)
-            self.reboot(mode)
+            try:
+                self.reboot(mode)
+            except RobotError as err:
+                self.server.logger.info("Reboot method %r failed, falling"
+                                        "back to %r reboot method...", tries[0], tries[1])
+                continue
```
Comment (aszlig) [inline hetzner/reset.py:93]: Apart from this possibly resulting in `IndexError`s, why is this needed?

**Review by aszlig — COMMENTED** [review]
Thanks a lot for your pull request. I do have one question though, please see the comment.

File: hetzner/reset.py (hunk 1 above)
Comment (acjohnson) [reply hetzner/reset.py:93]: Please explain where the risk of index error is. If you look at line 83 you'll see the `tries` list is hard coded, maybe you meant something else.

**Comment by acjohnson** [comment]
@aszlig thank you for taking the time to look at this. The reason for this is because when  I run `hetznerctl rescue <ip>` on an AX-41 I get the following `400`:
```shell
# hetznerctl rescue 135.181.6.108
Server #1244841: Tring to reboot using the 'soft' method.
Traceback (most recent call last):
  File "/home/acjohnson/Projects/hetzner-kube/venv/bin/hetznerctl", line 418, in <module>
    main()
  File "/home/acjohnson/Projects/hetzner-kube/venv/bin/hetznerctl", line 414, in main
    subcommand.execute(robot, parser, args)
  File "/home/acjohnson/Projects/hetzner-kube/venv/bin/hetznerctl", line 105, in execute
    server.rescue.shell(**kwargs)
  File "/home/acjohnson/Projects/hetzner-kube/venv/lib/python3.9/site-packages/hetzner/server.py", line 152, in shell
    self.observed_activate(*args, **kwargs)
  File "/home/acjohnson/Projects/hetzner-kube/venv/lib/python3.9/site-packages/hetzner/server.py", line 130, in observed_activate
    self.server.observed_reboot(*args, **kwargs)
  File "/home/acjohnson/Projects/hetzner-kube/venv/lib/python3.9/site-packages/hetzner/server.py", line 468, in observed_reboot
    return self.reset.observed_reboot(*args, **kwargs)
  File "/home/acjohnson/Projects/hetzner-kube/venv/lib/python3.9/site-packages/hetzner/reset.py", line 88, in observed_reboot
    self.reboot(mode)
  File "/home/acjohnson/Projects/hetzner-kube/venv/lib/python3.9/site-packages/hetzner/reset.py", line 145, in reboot
    return self.conn.post('/reset/{0}'.format(self.server.ip),
  File "/home/acjohnson/Projects/hetzner-kube/venv/lib/python3.9/site-packages/hetzner/robot.py", line 408, in post
    return self.request('POST', path, data)
  File "/home/acjohnson/Projects/hetzner-kube/venv/lib/python3.9/site-packages/hetzner/robot.py", line 402, in request
    raise RobotError(err, response.status)
hetzner.RobotError: 400 - invalid input, fields: type (400)
```
Also it would probably be a good idea to switch to using `server.number` for the `reset` calls as `server.ip` has been deprecated https://robot.your-server.de/doc/webservice/en.html#post-reset-server-number

File: hetzner/reset.py (hunk 1 above)
Comment (aszlig) [reply hetzner/reset.py:93]: Okay, maybe I'm blind then, but `observed_reboot(tries=['hard'])` should result in an `IndexError`, right?

**Comment by aszlig** [comment]
> ```python
>   File "/home/acjohnson/Projects/hetzner-kube/venv/lib/python3.9/site-packages/hetzner/reset.py", line 145, in reboot
>     return self.conn.post('/reset/{0}'.format(self.server.ip),
> ```

Isn't this what this PR is actually about?

**Comment by acjohnson** [comment]
> > ```python
> >   File "/home/acjohnson/Projects/hetzner-kube/venv/lib/python3.9/site-packages/hetzner/reset.py", line 145, in reboot
> >     return self.conn.post('/reset/{0}'.format(self.server.ip),
> > ```
> 
> Isn't this what this PR is actually about?

No deprecated doesn't mean removed, it just means it will be removed. The bug is that `sw` is not a valid reset type apparently.
```shell
# hetznerctl rescue <ip>
Server #1244841: Tring to reboot using the 'soft' method.
Traceback (most recent call last):
  File "/home/acjohnson/Projects/hetzner-kube/venv/bin/hetznerctl", line 418, in <module>
    main()
  File "/home/acjohnson/Projects/hetzner-kube/venv/bin/hetznerctl", line 414, in main
    subcommand.execute(robot, parser, args)
  File "/home/acjohnson/Projects/hetzner-kube/venv/bin/hetznerctl", line 105, in execute
    server.rescue.shell(**kwargs)
  File "/home/acjohnson/Projects/hetzner-kube/venv/lib/python3.9/site-packages/hetzner/server.py", line 152, in shell
    self.observed_activate(*args, **kwargs)
  File "/home/acjohnson/Projects/hetzner-kube/venv/lib/python3.9/site-packages/hetzner/server.py", line 130, in observed_activate
    self.server.observed_reboot(*args, **kwargs)
  File "/home/acjohnson/Projects/hetzner-kube/venv/lib/python3.9/site-packages/hetzner/server.py", line 468, in observed_reboot
    return self.reset.observed_reboot(*args, **kwargs)
  File "/home/acjohnson/Projects/hetzner-kube/venv/lib/python3.9/site-packages/hetzner/reset.py", line 88, in observed_reboot
    self.reboot(mode)
  File "/home/acjohnson/Projects/hetzner-kube/venv/lib/python3.9/site-packages/hetzner/reset.py", line 145, in reboot
    return self.conn.post('/reset/{0}'.format(self.server.number),
  File "/home/acjohnson/Projects/hetzner-kube/venv/lib/python3.9/site-packages/hetzner/robot.py", line 408, in post
    return self.request('POST', path, data)
  File "/home/acjohnson/Projects/hetzner-kube/venv/lib/python3.9/site-packages/hetzner/robot.py", line 402, in request
    raise RobotError(err, response.status)
hetzner.RobotError: 400 - invalid input, fields: type (400)
```

File: hetzner/reset.py (hunk 1 above)
Comment (acjohnson) [reply hetzner/reset.py:93]: The only way you'd get an index error is if the hard method failed, which I don't think that would likely happen.

If you prefer we can replace `tries[0]` `tries[1]` with strings.

Comment (aszlig) [reply hetzner/reset.py:93]: Well, or remove the `try`/`except` so that it is not needed in the first place, after all we already know the `_reset_types`, right? I'd like to have exceptions only be used for *exceptional* errors, aka errors we did **not** expect rather than make it part of the domain logic. We have enough of this in the code already and IMHO the goal should be to not introduce more of them.

Comment (acjohnson) [reply hetzner/reset.py]: The try/except is to catch the RobotError that I am receiving from `return self.conn.post('/reset/{0}'.format(self.server.number)` so unless that call will start working for AX-41 I don't know how else you propose we solve this. I went ahead and hard coded the strings instead.

**Comment by acjohnson** [comment]
Any more feedback @aszlig ?

File: hetzner/reset.py (hunk 1 above)
Comment (aszlig) [reply hetzner/reset.py]: We already know which reset types the server supports, so we don't need to run into an exception in the first place and also don't need to hardcode anything, right? As mentioned, adding logic based on exception is not what exceptions are meant for.

**Comment by aszlig** [comment]
> Any more feedback @aszlig ?

Yep, using the server number instead of the IP is fine, but I'm still not convinced with the `try...except`. If you can split those into different commits, I could cherry-pick the non-controversial parts.

File: hetzner/reset.py (hunk 1 above)
Comment (acjohnson) [reply hetzner/reset.py]: > We already know which reset types the server supports...
Great, in that case where can I get that info? We should just use that logic so that we don't hit this bug and then we won't need to catch the exception when it tries to send a soft reset and crash as it is doing currently

Comment (acjohnson) [reply hetzner/reset.py]: nvm, I think I see lemme take another stab at this...

**Comment by acjohnson** [comment]
@aszlig okay it's using `_reset_types` as previously suggested :sweat_smile:

**Comment by acjohnson** [comment]
@aszlig Happy new year! Lemme know if you think this will work now :smiley:

**Comment by acjohnson** [comment]
Rebased. @aszlig is this one mergeable as well? Thanks!

File: hetzner/reset.py [hunk 2]
```diff
@@ -76,14 +76,15 @@ def observed_reboot(self, patience=300, tries=None, manual=False):
         """
         is_down = False
 
+        self._update_status()
         if tries is None:
-            if self.server.is_vserver:
+            if self.server.is_vserver or 'sw' not in self._reset_types:
```
Comment (aszlig) [inline hetzner/reset.py:81]:
```suggestion
            if self.server.is_vserver or 'sw' not in self.reset_types:
```

File: hetzner/reset.py [hunk 3]
```diff
@@ -76,14 +76,15 @@ def observed_reboot(self, patience=300, tries=None, manual=False):
         """
         is_down = False
 
+        self._update_status()
```
Comment (aszlig) [inline hetzner/reset.py:79]:
```suggestion
```

File: hetzner/reset.py (hunk 2 above)
Comment (acjohnson) [reply hetzner/reset.py]: :facepalm:

File: hetzner/reset.py (hunk 3 above)
Comment (acjohnson) [reply hetzner/reset.py]: okay how 'bout now

**Comment by aszlig** [comment]
Hm, shouldn't `Reset._update_status` also use the number instead of the IP?

**Comment by acjohnson** [comment]
> Hm, shouldn't `Reset._update_status` also use the number instead of the IP?

Yeah and I'm not sure how this has evolved over the past several months...

According to https://robot.your-server.de/doc/webservice/en.html#get-reset-server-number it looks like server-number is preferred but the deprecation for server-ip is now crossed out lol

**Comment by acjohnson** [comment]
I went ahead and changed the GET to use server.number as well @aszlig

**Comment by acjohnson** [comment]
merge possible for this PR @aszlig?

**Comment by aszlig** [comment]
@acjohnson: I've cherry-picked ca5af16f87c6f05a6ceb08030a8bf20e154484ce for now, can you please rebase (and possibly fixup/squash your commits)? Alternatively, if you're fine with it[^1] I can do it as well.

[^1]: This is mainly because I wouldn't like it if someone would squash/fixup my own commits, but I can see that other people might not care about commit messages that much.

**Comment by acjohnson** [comment]
@aszlig done, look good? (I pushed the non-squashed/rebased branch to `rescuefixes-old` in my fork if you want to compare with the old branch)

File: hetzner/reset.py [hunk 4]
```diff
@@ -78,9 +78,13 @@ def observed_reboot(self, patience=300, tries=None, manual=False):
 
         if tries is None:
             tries = ['soft', 'hard']
+            if self.server.is_vserver or 'sw' not in self.reset_types:
+                tries = ['hard']
+            else:
+                tries = ['soft', 'hard']
```
Comment (aszlig) [inline hetzner/reset.py:84]: The `is_vserver` attribute no longer exists, do you mind dropping this hunk?

Comment (acjohnson) [reply hetzner/reset.py:84]: This won't work as I need to find a way to determine if `hard` reset_type should be used. Why does it default to `soft` when it is not supported by all server types (I'm using AX41 btw)? Here's the stacktrace I get if we can't use `hard` for the reset
```shell
# hetznerctl rescue -n -a <fingerprint> <server_ip>
Server #1245736: Trying to reboot using the 'soft' method.
Traceback (most recent call last):
  File "/home/ajohnson/Projects/aszlig-hetzner/venv/bin/hetznerctl", line 423, in <module>
    main()
  File "/home/ajohnson/Projects/aszlig-hetzner/venv/bin/hetznerctl", line 419, in main
    subcommand.execute(robot, parser, args)
  File "/home/ajohnson/Projects/aszlig-hetzner/venv/bin/hetznerctl", line 103, in execute
    server.rescue.observed_activate(**kwargs)
  File "/home/ajohnson/Projects/aszlig-hetzner/venv/lib/python3.10/site-packages/hetzner/server.py", line 130, in observed_activate
    self.server.observed_reboot(*args, **kwargs)
  File "/home/ajohnson/Projects/aszlig-hetzner/venv/lib/python3.10/site-packages/hetzner/server.py", line 467, in observed_reboot
    return self.reset.observed_reboot(*args, **kwargs)
  File "/home/ajohnson/Projects/aszlig-hetzner/venv/lib/python3.10/site-packages/hetzner/reset.py", line 85, in observed_reboot
    self.reboot(mode)
  File "/home/ajohnson/Projects/aszlig-hetzner/venv/lib/python3.10/site-packages/hetzner/reset.py", line 128, in reboot
    return self.conn.post('/reset/{0}'.format(self.server.number),
  File "/home/ajohnson/Projects/aszlig-hetzner/venv/lib/python3.10/site-packages/hetzner/robot.py", line 408, in post
    return self.request('POST', path, data)
  File "/home/ajohnson/Projects/aszlig-hetzner/venv/lib/python3.10/site-packages/hetzner/robot.py", line 402, in request
    raise RobotError(err, response.status)
hetzner.RobotError: 400 - invalid input, fields: type (400)

```

Comment (aszlig) [reply hetzner/reset.py:84]: Are you sure you're using this very branch? The `is_vserver` attribute [no longer exists](97d62b3ed131694956e721a591f09216f275eeb3), so you should get a different error.

Comment (acjohnson) [reply hetzner/reset.py:84]: Yeah I tested with removing lines 81-84, I don't mind you merging the `server.number` code, but that doesn't actually solve my problem which is the `soft` reset issue. :sweat_smile:

Comment (acjohnson) [reply hetzner/reset.py:84]: I'll go ahead and push the removal of 81-84 but I guess I'll have to do another PR to figure our the reset issue next. Thanks!
